soroban-test-wasms/wasm-workspace/linear_memory/src/lib.rs

```diff
@@ -51,4 +51,26 @@ impl Contract {
         assert!(e.vec_unpack_to_slice(vec, &mut long_buf).is_err());
         Ok(())
     }
+
+    // Bounce a map of vals off the host, successfully
+    pub fn map_mem_ok(e: Env) -> Result<(), Error> {
+        let map: MapObject = e.map_new_from_slices(&["a", "b"], &[1u32.into(), 2u32.into()])?;
+
+        let key_buf = ["a", "b"];
+        let mut val_buf: [Val; 2] = [Val::from(0); 2];
+
+        e.map_unpack_to_slice(map, &key_buf, &mut val_buf)?;
+        Ok(())
+    }
+
+    // Same but with out of order keys
+    pub fn map_mem_bad(e: Env) -> Result<(), Error> {
+        // out of order
+        assert!(e.map_new_from_slices(&["b", "a"], &[1u32.into(), 2u32.into()]).is_err());
```

Comment:
I agree with the assert part. The contract would've already trapped and the assert here does nothing (and the line after this will never get executed).
But I'm not sure about making `map_new_from_slices` and other functions infallible in the sdk. They are part of the `EnvBase` crate, and simply calls the host functions from the guest env. So I guess it doesn't really matter? If you have an opinion of what to do better in the sdk, can you create an issue there?

Reply:
> But I'm not sure about making map_new_from_slices and other functions infallible in the sdk. They are part of the EnvBase crate, and simply calls the host functions from the guest env. So I guess it doesn't really matter?

It does matter, because it creates a false impression that these functions can return an `Err` result. They cannot though, as they either cause a trap, or return `Ok`. This test is a live evidence of how misleading this is.

Reply:
I understand the issue with `EnvBase` though, so I'm not sure how exactly we should address this.

Reply:
I'm not totally familiar with the structuring of the sdk, maybe @graydon or @leighmcculloch can chime in here.

The `EnvBase` has same as the `guest::Env` here, where all functions have fallible interface but `Infallible` error. The `Env` functions are not directly exposed to the outside (unless via one of the public modules), but the `EnvBase` functions are. So maybe the answer here is not to expose the `EnvBase` methods, or expose only selected ones in other modules (and giving them infallible signatures)?

Reply:
> So maybe the answer here is not to expose the EnvBase methods, or expose only selected ones in other modules (and giving them infallible signatures)?

Yes, probably the least confusing thing to do would be to add a public wrapper over all the `EnvBase` functions and hide the fallible implementation, so that it can't be used by accident.

Reply:
Opened https://github.com/stellar/rs-soroban-sdk/issues/1169